crates/wysiwyg/src/composer_model/replace_text.rs

```diff
@@ -63,50 +63,65 @@ where
-                self.state.dom.find_parent_list_item_or_self(handle);
-            if let Some(parent_list_item_handle) = parent_list_item_handle {
-                let list_item_end_offset = range
-                    .locations
-                    .into_iter()
-                    .filter(|loc| loc.kind == ListItem)
-                    .next()
-                    .unwrap()
-                    .end_offset;
-                self.do_enter_in_list(
-                    &parent_list_item_handle,
-                    current_cursor_global_location,
-                    list_item_end_offset,
-                )
-            } else {
-                self.do_enter_in_text(handle, location.start_offset)
+        let position = range.start();
+        let leaf_at_cursor: Option<&DomLocation> = range.leaves().find(|loc| {
+            loc.position <= position && position <= loc.position + loc.length
+        });
+
+        match leaf_at_cursor {
+            None => {
+                // Selection doesn't contain any text like nodes. We can assume it's an empty Dom.
+                self.state
+                    .dom
+                    .document_mut()
+                    .append_child(DomNode::new_line_break());
+                self.state.start += 1;
+                self.state.end = self.state.start;
+                return self.create_update_replace_all();
+            }
+            Some(leaf) => {
+                if leaf.length == 0 {
+                    // Special case, there is an empty text node at the cursor position
+                    self.enter_with_zero_length_selection_and_empty_text_nodes(
+                        range,
+                    );
+                    return self.create_update_replace_all();
+                } else {
+                    let current_cursor_global_location =
+                        leaf.position + leaf.start_offset;
+                    let handle = &leaf.node_handle;
+                    let parent_list_item_handle =
+                        self.state.dom.find_parent_list_item_or_self(handle);
+
+                    match parent_list_item_handle {
+                        Some(handle) => {
+                            let list_item_end_offset = range
+                                .locations
+                                .into_iter()
+                                .filter(|loc| loc.kind == ListItem)
+                                .next()
+                                .unwrap()
+                                .end_offset;
+                            self.do_enter_in_list(
+                                &handle,
+                                current_cursor_global_location,
+                                list_item_end_offset,
+                            )
+                        }
+                        None => {
+                            self.do_enter_in_text(handle, leaf.start_offset)
+                        }
+                    }
+                }
```

Comment:
This is the same branching logic as before, but now we need to account for multiple locations so can't use the length check we were previously using